File: src/v5/EmployeeDbControl.java
package v5;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Scanner;

public class EmployeeDbControl {
    private final Connection connection;

    public EmployeeDbControl(Connection connection) {
        this.connection = connection;
    }

    //Method to create new Employee into the database
    public void addEmployeeToDatabase(Employee employee) {
        String sql = "INSERT INTO employee (fName, nickname, gender, phoneNumber, address, email, type, salary) VALUES (?, ?, ?, ?, ?, ?, ?, ?)";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, employee.getName());
            preparedStatement.setString(2, employee.getNickname());
            preparedStatement.setString(3, employee.getGender());
            preparedStatement.setString(4, employee.getPhoneNumber());
            preparedStatement.setString(5, employee.getAddress());
            preparedStatement.setString(6, employee.getEmail());
            preparedStatement.setString(7, employee.getEmployeeType());
            preparedStatement.setDouble(8, employee.getSalary());


            preparedStatement.executeUpdate();
            System.out.println("Employee added successfully to the database!");
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // Method to display table
    private static void displayTable(String[] data) {
        int numColumns = data.length;

        // Display table content
        int numRows = (int) Math.ceil((double) data.length / numColumns);
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < numColumns; j++) {
                int index = i + j * numRows;
                if (index < data.length) {
                    System.out.printf("%-20s", data[index]);
                }
            }
            System.out.println();
        }
    }


    // Method to retrieve data as to table
    public void viewEmployeesFromDatabase() {
        String sql = "SELECT idEmployee, fName, nickname, phoneNumber, type, salary FROM employee";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql);
             ResultSet resultSet = preparedStatement.executeQuery()) {

            if (!resultSet.isBeforeFirst()) {
                System.out.println("No Employees found in the database.");
            } else {
                System.out.println("Employees in the database:");

                // Table headers
                String[] headers = {"ID", "Name", "Phone Number", "Employee Type", "Salary"};
                displayTable(headers);

                while (resultSet.next()) {
                    int employeeId = resultSet.getInt("idEmployee");
                    String firstName = resultSet.getString("fName");
                    String lastName = resultSet.getString("nickName");
                    String phoneNumber = resultSet.getString("phoneNumber");
                    String employeeType = resultSet.getString("type");
                    double salary = resultSet.getDouble("salary");

                    // Formatted row data
                    String[] rowData = {
                            String.valueOf(employeeId),
                            firstName + " " + lastName,
                            phoneNumber,
                            employeeType,
                            String.valueOf(salary)
                    };

                    displayTable(rowData);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    //method to delete employee in the database
    public void deleteEmployeeFromDatabase(int employeeId) {
        String sql = "DELETE FROM employee WHERE idEmployee = ?";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, employeeId);

            int rowsAffected = preparedStatement.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("Employee with ID " + employeeId + " deleted successfully.");
            } else {
                System.out.println("No Employee found with ID " + employeeId + ". No deletion performed.");
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    //method to Update
    public void updateEmployeeInfo(int employeeId, String columnName, String newValue) {
        String sql = "UPDATE Employee SET " + columnName + " = ? WHERE idEmployee = ?";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, newValue);
            preparedStatement.setInt(2, employeeId);

            int rowsAffected = preparedStatement.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("Employee with ID " + employeeId + " updated successfully.");
            } else {
                System.out.println("No Employee found with ID " + employeeId + ". No update performed.");
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    //DISPLAY
    public void displayColumnTable() {
        String[] columnNames = {
                "Name", "Nickname", "Gender", "Phone Number",
                "Address", "Email", "Subscription Type",
                "Ampere", "Status", "Balance"
        };

        System.out.println("Column Index | Column Name");
        System.out.println("-------------|-------------");
        for (int i = 0; i < columnNames.length; i++) {
            System.out.printf("%12d | %s\n", i + 1, columnNames[i]);
        }
    }

    // Method to get user input for ID, column index, and new value
    public void getUserInputAndUpdate() {
        Scanner scanner = new Scanner(System.in);
        EmployeeDbControl custDbControl = new EmployeeDbControl(connection);

        // Display the table of column names with index numbers
        displayColumnTable();

        // Get user input for ID, column index, and new value
        System.out.print("\nEnter Employee ID: ");
        int employeeId = scanner.nextInt();

        System.out.print("Enter column index to update: ");
        int columnIndex = scanner.nextInt();

        scanner.nextLine(); // Consume newline left from nextInt()

        System.out.print("Enter new value: ");
        String newValue = scanner.nextLine();

        // Call the update method with provided details
        custDbControl.updateEmployeeInfo(employeeId, getColumnName(columnIndex), newValue);
    }

    // Method to retrieve column name based on index
    private String getColumnName(int index) {
        String[] columnNames = {
                "fName", "nickname", "gender", "phoneNumber",
                "address", "email", "subscriptionType",
                "ampere", "status", "balance"
        };
        if (index > 0 && index <= columnNames.length) {
            return columnNames[index - 1];
        }
        return "";
    }

}

File: pythonCodeConversion/electricbill/controllers/employee_db_control.py
import mysql.connector


class EmployeeDbControl:
    def __init__(self, connection):
        self.connection = connection

    def add_employee_to_database(self, employee):
        sql = "INSERT INTO employee (fName, nickname, gender, phoneNumber, address, email, type, salary) VALUES (%s, " \
              "%s, %s, %s, %s, %s, %s, %s)"

        try:
            cursor = self.connection.cursor()
            cursor.execute(sql, (employee.get_name(), employee.get_nickname(), employee.get_gender(),
                                 employee.get_phone_number(), employee.get_address(), employee.get_email(),
                                 employee.get_employee_type(), employee.get_salary()))
            self.connection.commit()
            print("Employee added successfully to the database!")
        except mysql.connector.Error as e:
            print("Error adding employee to the database:", e)
        finally:
            cursor.close()

    @staticmethod
    def display_table(data):
        num_columns = len(data)
        num_rows = (num_columns + 1) // 2

        for i in range(num_rows):
            for j in range(2):
                index = i + j * num_rows
                if index < num_columns:
                    print(f"{data[index]:<20}", end="")
            print()

    def view_employees_from_database(self):
        sql = "SELECT idEmployee, fName, nickname, phoneNumber, type, salary FROM employee"

        try:
            cursor = self.connection.cursor()
            cursor.execute(sql)
            result = cursor.fetchall()

            if not result:
                print("No Employees found in the database.")
            else:
                print("Employees in the database:")
                headers = ["ID", "Name", "Phone Number", "Employee Type", "Salary"]
                self.display_table(headers)

                for row in result:
                    employee_id, first_name, last_name, phone_number, employee_type, salary = row
                    display_name = f"{first_name} {last_name}"

                    row_data = [str(employee_id), display_name, phone_number, employee_type, str(salary)]
                    self.display_table(row_data)
        except mysql.connector.Error as e:
            print("Error viewing employees from the database:", e)
        finally:
            cursor.close()

    def delete_employee_from_database(self, employee_id):
        sql = "DELETE FROM employee WHERE idEmployee = %s"

        try:
            cursor = self.connection.cursor()
            cursor.execute(sql, (employee_id,))
            self.connection.commit()
            if cursor.rowcount > 0:
                print(f"Employee with ID {employee_id} deleted successfully.")
            else:
                print(f"No Employee found with ID {employee_id}. No deletion performed.")
        except mysql.connector.Error as e:
            print("Error deleting employee from the database:", e)
        finally:
            cursor.close()

    def update_employee_info(self, employee_id, column_name, new_value):
        sql = f"UPDATE employee SET {column_name} = %s WHERE idEmployee = %s"

        try:
            cursor = self.connection.cursor()
            cursor.execute(sql, (new_value, employee_id))
            self.connection.commit()
            if cursor.rowcount > 0:
                print(f"Employee with ID {employee_id} updated successfully.")
            else:
                print(f"No Employee found with ID {employee_id}. No update performed.")
        except mysql.connector.Error as e:
            print("Error updating employee info:", e)
        finally:
            cursor.close()

    @staticmethod
    def display_column_table():
        column_names = [
            "Name", "Nickname", "Gender", "Phone Number",
            "Address", "Email", "Employee Type",
            "Salary"
        ]

        print("Column Index | Column Name")
        print("-------------|-------------")
        for i, column_name in enumerate(column_names, start=1):
            print(f"{i:<12} | {column_name}")

    def get_user_input_and_update(self):
        self.display_column_table()

        employee_id = int(input("\nEnter Employee ID: "))
        column_index = int(input("Enter column index to update: "))
        new_value = input("Enter new value: ")

        column_name = self.get_column_name(column_index)
        self.update_employee_info(employee_id, column_name, new_value)

    @staticmethod
    def get_column_name(index):
        column_names = [
            "fName", "nickname", "gender", "phoneNumber",
            "address", "email", "type",
            "salary"
        ]
        if 0 < index <= len(column_names):
            return column_names[index - 1]
        return ""
